syntax = "proto3";

package navigation;

import "google/protobuf/empty.proto";

message Descartes {
    float x = 1;		// 米
    float y = 2;		// 米
    float z = 3;
}

message Euler {
    float roll = 1;
    float pitch = 2;
    float yaw = 3;		// 弧度
}

message Pose {
    Descartes position = 1;	// 位置
    Euler attitude = 2;		// 姿态
}

message NaviState {
    Descartes position = 1;	// 位置
    Descartes velocity = 2;	// 速度
    Euler attitude = 3;		// 姿态
    bool navigating = 4;	// 是否正在导航
}

message Response {
    bool succeeded = 1;
    string msg = 2;
}

message NaviResponse {
    bool succeeded = 1;
    string msg = 2;
    bool arrived = 3;		// 是否已到达
    NaviState state = 4;
}

message Config {
    bool relative = 1;		// 相对/绝对目标位姿
}

message Map {
    Descartes origin = 1;		// 原点坐标
    float resolution = 2;		// 分辨率
    float occupied_threshold = 3;	// 占据阀阈
    float free_threshold = 4;		// 空置阀阈
    bytes pgm = 5;			// PGM格式地图图片
}

service NaviController {
    rpc setDestination(Pose) returns (Response);
    rpc startNavi(Config) returns (stream NaviResponse);
    rpc stopNavi(google.protobuf.Empty) returns (Response);
    rpc getState(google.protobuf.Empty) returns (NaviState);
}

service MapManager {
    rpc getMap(google.protobuf.Empty) returns (Map);
}
